pkg/chunk/local/fs_object_client.go

```diff
@@ -127,12 +134,28 @@ func (f *FSObjectClient) List(ctx context.Context, prefix string) ([]chunk.Stora
 }
 
 func (f *FSObjectClient) DeleteObject(ctx context.Context, objectKey string) error {
-	err := os.Remove(path.Join(f.cfg.Directory, objectKey))
-	if err != nil && os.IsNotExist(err) {
-		return chunk.ErrStorageObjectNotFound
+	filePath := filepath.Join(f.cfg.Directory, objectKey)
+
+	err := os.Remove(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return chunk.ErrStorageObjectNotFound
+		}
+		return err
+	}
+
+	// remove the parent directory when it is empty
+	parentDir := filepath.Dir(filePath)
+	ok, err := isDirEmpty(parentDir)
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		return nil
 	}
 
-	return err
+	return os.RemoveAll(parentDir)
```

Comment:
Using `RemoveAll()` may suffer race conditions. If another process has created an object in the meanwhile (after your call to `isDirEmpty()`, this `RemoveAll()` call will erroneously delete it too. You should use `Remove()` here and check if the returned error is "directory not empty" and, if so, ignore it (because it's OK).